### PR TITLE
Initial implementation of requestIdleCallback on iOS

### PR DESCRIPTION
**Pull request by janicduplessis**
iOS follow up to #8569. This currently depends on the Android PR since it contains the JS implementation, only review the last commit. Just putting this out here for visibility, don't merge this before the Android PR.

**Test plan**
Tested by running a background task that burns all remaining idle time (see UIExplorer example).

Tested that native only calls into JS when there are pending idle callbacks.

Tested that timers are executed before idle callback.

### CODE REVIEW

**Comment by facebook-github-bot-0** [comment]
By analyzing the blame information on this pull request, we identified @javache and @yungsters to be potential reviewers.

**Comment by janicduplessis** [comment]
cc @brentvatne @ide

File: React/Modules/RCTTiming.m [hunk 1]
```diff
@@ -181,22 +192,31 @@ - (void)didUpdateFrame:(__unused RCTFrameUpdate *)update
...
-    // the displaylink frivolously.
-    return;
   }
 
-  // No need to call the pauseCallback as RCTDisplayLink will ask us about our paused
-  // status immediately after completing this call
-  if (_timers.count == 0) {
-    _paused = YES;
+  if (_sendIdleEvents) {
+    NSTimeInterval frameElapsed = (CACurrentMediaTime() - update.timestamp) * 1000;
```
Comment (janicduplessis) [inline React/Modules/RCTTiming.m]: Not sure if there is a better way to convert this value to a unix timestamp, sometimes it is not in sync with the JS `Date.now` as I get some frame time left greater that 17ms (varies from 15-17.5).

`CACurrentMediaTime() - update.timestamp` is always positive so this is not the issue. I suspect that `[[NSDate date] timeIntervalSince1970]` gets out of sync with JS `Date.now`.

File: React/Modules/RCTTiming.m [hunk 2]
```diff
@@ -181,22 +192,31 @@ - (void)didUpdateFrame:(__unused RCTFrameUpdate *)update
...
+    NSTimeInterval currentTimestamp = [[NSDate date] timeIntervalSince1970] * 1000;
+    NSNumber *absoluteFrameStartMillis = [NSNumber numberWithDouble:currentTimestamp - frameElapsed];
```
Comment (ide) [inline React/Modules/RCTTiming.m]: Move this into the if-clause so we avoid an extra object allocation when there is not enough remaining time, and use `@()` syntax

File: React/Modules/RCTTiming.m [hunk 3]
```diff
@@ -133,6 +136,14 @@ - (dispatch_queue_t)methodQueue
   return RCTJSThread;
 }
 
+- (NSDictionary *)constantsToExport
+{
+  return @{
+    @"frameDuration": [NSNumber numberWithDouble:kFrameDuration],
```
Comment (ide) [inline React/Modules/RCTTiming.m]: Syntax: `@(kFrameDuration)` should work too

File: React/Modules/RCTTiming.m [hunk 4]
```diff
@@ -181,22 +192,31 @@ - (void)didUpdateFrame:(__unused RCTFrameUpdate *)update
...
+    if (kFrameDuration - frameElapsed >= kIdleCallbackFrameDeadline) {
+      [_bridge enqueueJSCall:@"JSTimersExecution.callIdleCallbacks" args:@[absoluteFrameStartMillis]];
+    }
   }
-  // If the next timer is more than 1 second out, pause and schedule an NSTimer;
-  else if ([nextScheduledTarget timeIntervalSinceNow] > kMinimumSleepInterval) {
-    [self scheduleSleepTimer:nextScheduledTarget];
-    _paused = YES;
+
+  // If we call at least one timer this frame, don't switch to a paused state yet, so if
```
Comment (ide) [inline React/Modules/RCTTiming.m]: Since the if-statement's logic was reversed, maybe rephrase this as:

> Switch to a paused state *only* if we didn't call at least one timer this frame, so if...

File: React/Modules/RCTTiming.m [hunk 5]
```diff
@@ -181,22 +192,31 @@ - (void)didUpdateFrame:(__unused RCTFrameUpdate *)update
...
+  // in response to this timer another timer is scheduled, we don't pause and unpause
+  // the displaylink frivolously.
+  if (!_sendIdleEvents && timersToCall.count == 0) {
+    // No need to call the pauseCallback as RCTDisplayLink will ask us about our paused
+    // status immediately after completing this call
+    if (_timers.count == 0) {
+      _paused = YES;
+    }
+    // If the next timer is more than 1 second out, pause and schedule an NSTimer;
+    else if (!_sendIdleEvents && [nextScheduledTarget timeIntervalSinceNow] > kMinimumSleepInterval) {
```
Comment (ide) [inline React/Modules/RCTTiming.m]: Can remove the `!_sendIdleEvents` check since it's always true due to the outer if-statement

File: React/Modules/RCTTiming.m [hunk 6]
```diff
@@ -151,7 +162,7 @@ - (void)stopTimers
 
 - (void)startTimers
 {
-  if (!_bridge || _timers.count == 0) {
+  if (!_bridge || (!_sendIdleEvents && _timers.count == 0)) {
```
Comment (ide) [inline React/Modules/RCTTiming.m]: Might want to define a little helper method that returns `!_sendIdleEvents && _timers.count == 0`

File: React/Modules/RCTTiming.m (hunk 2 above)
Comment (ide) [inline React/Modules/RCTTiming.m]: Also it's not a big deal but might be clearer to do the x1000 multiplication at the very end, so leave `frameElapsed` and `currentTimestamp` in seconds (which NSTimeInterval is supposed to be)

Comment (janicduplessis) [inline React/Modules/RCTTiming.m]: Should I change the frameDuration and frameDeadline constants to seconds then? And multiply it by 1000 before sending it to JS.

Comment (ide) [inline React/Modules/RCTTiming.m]: Yeah. The way I think of this is: when working in Obj-C/Cocoa we want to follow Cocoa conventions most of the time (in this case, use `NSTimeInterval` and define the values in seconds, which is what people expect NSTimeInterval to represent). And then, as late as possible, convert from seconds to milliseconds when we are sending these values from Obj-C to JS.

File: React/Modules/RCTTiming.m (hunk 1 above)
Comment (ide) [inline React/Modules/RCTTiming.m]: We could synchronize the clocks on each side by doing a super-fast, synchronous call from JS -> Obj-C to compute the clock skew. Once we know the clock skew, it's easy to convert Obj-C timestamps to JS timestamps. RN might also already be tracking the clock skew deep within the bridge and message queue. If so we could maybe piggyback off of that.

**Comment by ide** [comment]
LGTM, will wait for Android to land.

**Comment by ide** [comment]
@janicduplessis can you rebase this on top of master now that rIC for Android has landed?

**Comment by janicduplessis** [comment]
@ide Done!

**Comment by ide** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-2** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/602178519952306/int_phab) to review.

File: React/Modules/RCTTiming.m [hunk 7]
```diff
@@ -133,6 +138,14 @@ - (dispatch_queue_t)methodQueue
   return RCTJSThread;
 }
 
+- (NSDictionary *)constantsToExport
```
Comment (javache) [inline React/Modules/RCTTiming.m]: Please define this just on the JS side. This will require the module to be created on the main thread just to send some static numbers over.

File: React/Modules/RCTTiming.m [hunk 8]
```diff
@@ -163,6 +176,11 @@ - (void)startTimers
   }
 }
 
+- (BOOL)hasPendingTimers
+{
+  return _sendIdleEvents || _timers.count > 0;
+}
+
 - (void)didUpdateFrame:(__unused RCTFrameUpdate *)update
```
Comment (javache) [inline React/Modules/RCTTiming.m]: So this will get called at the beginning I believe of every frame, I don't think this does what you want it to do in terms of sending idle events.

**Comment by javache** [comment]
I just got an internal task that this caused a perf regression. I'm still investigating if it's legit, but for any significant changes to timers could you make sure someone from FB has a look as well?

File: React/Modules/RCTTiming.m (hunk 8 above)
Comment (janicduplessis) [inline React/Modules/RCTTiming.m]: I think this is what we want in this case, requestIdleCallback is called at the end of every frame that has some time left with the frame start time. Then JS can schedule work knowing how much time it has left.

File: React/Modules/RCTTiming.m (hunk 7 above)
Comment (ide) [inline React/Modules/RCTTiming.m]: We should add comments on both sides saying to keep the constants in sync across JS and Obj-C then

**Comment by ide** [comment]
It's strange that this caused a regression, maybe it has to do with initializing the module on the main thread? If no one has called rIC from JS, the code that runs on each frame shouldn't do any more work aside from an extra boolean ivar check.
